File: DS_P1_S3/C++/CompruebaRentabilidad.cpp
#include "CompruebaRentabilidad.h"

FormularioPrestamo* CompruebaRentabilidad::ejecutar(FormularioPrestamo* f) {
    bool rentable = false;
    string opcion;

    cout << "Comprobando la rentabilidad del cliente ..." << endl << endl;
    if (comprueba_rentabilidad(f) == false){
        cout << "En base a su solicitud el banco le ofrece los siguientes términos." << endl;
        //cin >> opcion;

        //  Toma de decisiones aleatoria
        vector<string> opciones = {"si", "no"};
        srand (time(NULL));
        string opcion =  opciones[(rand() % opciones.size())];

        if(opcion == "si"){
            cout << "Usted acepta las condiciones, asignando prestamo..." << endl;
        }
        else{
            cout << "Como usted no aceptó las condiciones estamos cancelando su solicitud..." << endl;
            f->setInteres(NULL);
            f->setCuantia(NULL);
            f->setCuotas(NULL);
            f->setCuenta(NULL);
        }
    }
    return f;
}

bool CompruebaRentabilidad::comprueba_rentabilidad(FormularioPrestamo* f) {
    bool no_rentable = true;
    float tipo = f->getInteres()/1200;

    double numerador = tipo * pow(1+ tipo, f->getCuotas());
    double denominador = pow(1 + tipo, f->getCuotas()) - 1;
    double cuota = f->getCuantia() * (numerador / denominador);

    //cout << "Calculando cuota teórica" << endl;
    //cout << "Prestamo de " << f->getCuantia() << "€ Cuota teórica " << cuota << " €/mes en " << f->getCuotas() << " meses, nómina cliente es " << f->getCuenta()->getNomina() << endl;
    while(no_rentable){

        numerador = tipo * pow(1+ tipo, f->getCuotas());
        denominador = pow(1 + tipo, f->getCuotas()) - 1;
        cuota = f->getCuantia() * (numerador / denominador);

        // Si supera la nómina mensual del cliente
        if(cuota > f->getCuenta()->getNomina()*0.6){
            // Aplicamos el filtro de rentabilidad para el importe solicitado
            f->setCuotas(f->getCuotas()+1);
        }
        else {
            no_rentable = false;
            cout << "El banco le ofrece un prestamo de " << f->getCuantia() << " € su cuota seria " << cuota << " €/mes en " << f->getCuotas() << " meses con un interés del " << f->getInteres() << endl;
        }
    }
    return no_rentable;
}

File: DS_P1_S3/Dart/p1_s3/lib/CompruebaRentabilidad.dart
import 'dart:convert';
import 'dart:math';
import 'Filtro.dart';
import 'FormularioPrestamo.dart';
import 'dart:io';

class CompruebaRentabilidad implements Filtro{
  FormularioPrestamo ejecutar(FormularioPrestamo f){
    bool rentable = false;
    String opcion;

    print('Comprobando la rentabilidad del cliente ...\n');
    if (comprueba_rentabilidad(f) == false){
      print('En base a su solicitud el banco le ofrece los siguientes términos.');
    //  var opcion = stdin.readLineSync(encoding: Encoding.getByName('utf-8'));   // Para leer de entrada

      //  Toma de decisión aleatoria
      List<String> opciones = ['si', 'no'];
      Random random = new Random();
      var opcion = opciones[random.nextInt(opciones.length)];

      if(opcion == 'S' || opcion == 's' || opcion == 'si'){
        print('Usted acepta las condiciones, asignando prestamo...');
      }
      else{
        print('Como usted no aceptó las condiciones estamos cancelando su solicitud...');
        f.interes = null;
        f.cuantia = null;
        f.cuotas = null;
        f.cuenta = null;
      }
    }
    return f;
  }
  bool comprueba_rentabilidad(FormularioPrestamo f){
    bool no_rentable = true;
    double tipo = f.interes/1200;

    double numerador = tipo * pow(1+ tipo, f.cuotas);
    double denominador = pow(1 + tipo, f.cuotas) - 1;
    double cuota = f.cuantia * (numerador / denominador);

    //print('Calculando cuota teórica\n');
    //print('Prestamo de ${f.cuantia.toStringAsFixed(2)} € Cuota teórica ${cuota.toStringAsFixed(2)} €/mes en ${f.cuotas} meses, nómina cliente es ${f.cuenta.nomina.toStringAsFixed(2)}');
    while(no_rentable){

      numerador = tipo * pow(1+ tipo, f.cuotas);
      denominador = pow(1 + tipo, f.cuotas) - 1;
      cuota = f.cuantia * (numerador / denominador);

      // Si supera la nómina mensual del cliente
      if(cuota > f.cuenta.nomina*0.6){
        // Aplicamos el filtro de rentabilidad para el importe solicitado
        f.cuotas++;
      }
      else {
        no_rentable = false;
        print('El banco le ofrece un prestamo de ${f.cuantia.toStringAsFixed(2)} € su cuota seria ${cuota.toStringAsFixed(2)} €/mes en ${f.cuotas} meses con un interés del ${f.interes.toStringAsFixed(2)}');
        //print('Es rentable');
      }
    }
    return no_rentable;
  }
}